File: Ass_6_11_1.py
import sys

# Increase recursion depth just in case the graph is huge
sys.setrecursionlimit(200000)

def dfs(node, adj, visited):
    visited[node] = True
    
    # Visit all connected neighbors
    for neighbor in adj[node]:
        if not visited[neighbor]:
            dfs(neighbor, adj, visited)

def solve():
    try:
        # Read directly from input.txt 
        input_data = open("input.txt").read().split()
    except FileNotFoundError:
        print("Please create input.txt with your input data.")
        return

    if not input_data:
        return
        
    n = int(input_data[0])
    m = int(input_data[1])
    
    # 1. Build the Graph (Adjacency List)
    adj = [[] for _ in range(n)]
    
    idx = 2
    for _ in range(m):
        u = int(input_data[idx])
        v = int(input_data[idx+1])
        adj[u].append(v)
        adj[v].append(u)
        idx += 2
        
    # 2. Track visited nodes
    visited = [False] * n
    components = 0
    
    # 3. Count Components using DFS
    for i in range(n):
        if not visited[i]:
            components += 1
            dfs(i, adj, visited)
            
    print(components)

if __name__ == "__main__":
    solve()

File: Ass_6_11_1.java
import java.util.*;

public class Ass_6_11_1 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        
        if (sc.hasNextInt()) {
            int N = sc.nextInt(); // Number of nodes/computers
            int M = sc.nextInt(); // Number of edges/connections
            
            // 1. Build the Graph (Adjacency List)
            List<List<Integer>> adj = new ArrayList<>();
            for (int i = 0; i < N; i++) {
                adj.add(new ArrayList<>());
            }
            
            for (int i = 0; i < M; i++) {
                int u = sc.nextInt();
                int v = sc.nextInt();
                // Since it's an undirected network, connect both ways
                adj.get(u).add(v);
                adj.get(v).add(u);
            }
            
            // 2. Track visited nodes
            boolean[] visited = new boolean[N];
            int components = 0;
            
            // 3. Count Components using DFS
            for (int i = 0; i < N; i++) {
                if (!visited[i]) {
                    components++; // Found a new isolated subnetwork
                    dfs(i, adj, visited); // Mark all connected computers as visited
                }
            }
            
            System.out.println(components);
        }
        sc.close();
    }
    
    // Recursive DFS function
    public static void dfs(int node, List<List<Integer>> adj, boolean[] visited) {
        visited[node] = true; // Mark current computer as visited
        
        // Visit all its neighbors
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                dfs(neighbor, adj, visited);
            }
        }
    }
}